hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/DataSerializableConventionsTest.java

```diff
@@ -0,0 +1,222 @@
+        }
+    }
+
+    // fails when IdentifiedDataSerializable classes:
+    // - do not have a default no-args constructor
+    // - factoryId/id pairs are not unique per class
+    @Test
+    public void test_identifiedDataSerializables_haveUniqueFactoryAndTypeId()
+            throws IllegalAccessException, InvocationTargetException, ClassNotFoundException {
+        Set<String> classesWithInstantiationProblems = new TreeSet<String>();
+        Set<String> classesThrowingUnsupportedOperationException = new TreeSet<String>();
+
+        Multimap<Integer, Integer> factoryToTypeId = HashMultimap.create();
+
+        Set<Class> identifiedDataSerializables = getIDSProductionConcreteClasses();
+
+        for (Class<? extends IdentifiedDataSerializable> klass : identifiedDataSerializables) {
+            // exclude classes which are known to be meant for local use only
+            if (!AbstractLocalOperation.class.isAssignableFrom(klass)) {
+                // wrap all of this in try-catch, as it is legitimate for some classes to throw UnsupportedOperationException
+                try {
+                    Constructor<? extends IdentifiedDataSerializable> ctor = klass.getDeclaredConstructor();
+                    ctor.setAccessible(true);
+                    IdentifiedDataSerializable instance = ctor.newInstance();
+                    int factoryId = instance.getFactoryId();
+                    int typeId = instance.getId();
+                    if (factoryToTypeId.containsEntry(factoryId, typeId)) {
+                        fail("Factory-Type ID pair {" + factoryId + ", " + typeId + "} from " + klass.toString() + " is already "
+                                + "registered in another type.");
+                    } else {
+                        factoryToTypeId.put(factoryId, typeId);
+                    }
+                } catch (UnsupportedOperationException e) {
+                    // expected from local operation classes not meant for serialization
+                    // gather those and print them to system.out for information at end of test
+                    classesThrowingUnsupportedOperationException.add(klass.getName());
+                } catch (InstantiationException e) {
+                    classesWithInstantiationProblems.add(klass.getName() + " failed with " + e.getMessage());
+                } catch (NoSuchMethodException e) {
+                    classesWithInstantiationProblems.add(klass.getName() + " failed with " + e.getMessage());
+                }
+            }
+        }
+
+        if (!classesThrowingUnsupportedOperationException.isEmpty()) {
+            System.out.println("INFO: " + classesThrowingUnsupportedOperationException.size() + " classes threw "
+                    + "UnsupportedOperationException in getFactoryId/getId invocation:");
+            for (String className : classesThrowingUnsupportedOperationException) {
+                System.out.println(className);
```

Comment:
An `IdentifiedDataSerializable` implementation throwing `UnsupportedOperationException` from `getFactoryId` and/or `getId` is expected (eg local-only operations such as `FinalizeMigrationOperation` which are not supposed to be deserialized/serialized over the network). So this is intended to be only reported for information on the test's output, it should not fail the test.